src/IO/ZstdDeflatingWriteBuffer.cpp

```diff
@@ -0,0 +1,94 @@
+#include <IO/ZstdDeflatingWriteBuffer.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int ZSTD_ENCODER_FAILED;
+}
+
+ZstdDeflatingWriteBuffer::ZstdDeflatingWriteBuffer(
+    std::unique_ptr<WriteBuffer> out_, int compression_level, size_t buf_size, char * existing_memory, size_t alignment)
+    : BufferWithOwnMemory<WriteBuffer>(buf_size, existing_memory, alignment), out(std::move(out_))
+{
+    cctx = ZSTD_createCCtx();
+    if (cctx == nullptr)
+        throw Exception(ErrorCodes::ZSTD_ENCODER_FAILED, "zstd stream encoder init failed: zstd version: {}", ZSTD_VERSION_STRING);
+    ZSTD_CCtx_setParameter(cctx, ZSTD_c_compressionLevel, compression_level);
+    ZSTD_CCtx_setParameter(cctx, ZSTD_c_checksumFlag, 1);
+
+    input = {nullptr, 0, 0};
+    output = {nullptr, 0, 0};
+}
+
+
+ZstdDeflatingWriteBuffer::~ZstdDeflatingWriteBuffer()
+{
+    try
+    {
+        finish();
+
+        ZSTD_freeCCtx(cctx);
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+}
+
+void ZstdDeflatingWriteBuffer::nextImpl()
+{
+    if (!offset())
+        return;
+
+    bool last_chunk = hasPendingData();
```

Comment:
Difficult to understand the meaning of this variable.
How it's possible that "if there is more data it is the last chunk"?

Reply:
Do we have a test with more than 1MB of data?